core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildRemoteContainerRunner.java

```diff
@@ -34,59 +39,93 @@ protected void preBuild(Path outputDir, List<String> buildArgs) throws Interrupt
+                containerRuntime.getExecutableName(), "cp", outputDir.toAbsolutePath() + "/.",
                 containerId + ":" + NativeImageBuildStep.CONTAINER_BUILD_VOLUME_PATH };
-        runCommand(copyCommand, "Failed to copy source-jar and libs from host to builder container", null);
+        runCommand(copyCommand, "Failed to copy source-jar and libs from host to builder container");
         super.preBuild(outputDir, buildArgs);
     }
 
-    private String runCommandAndReadOutput(String[] command, String errorMsg) throws IOException, InterruptedException {
+    private List<String> runCommandAndReadOutput(String[] command) throws IOException, InterruptedException {
         log.info(String.join(" ", command).replace("$", "\\$"));
-        Process process = new ProcessBuilder(command).start();
+        final Process process = new ProcessBuilder(command).start();
         if (process.waitFor() != 0) {
-            throw new RuntimeException(errorMsg);
+            throw new RuntimeException("Command failed: " + String.join(" ", command));
         }
         try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
-            return reader.readLine();
+            return reader.lines().toList();
         }
     }
 
     @Override
     protected void postBuild(Path outputDir, String nativeImageName, String resultingExecutableName) {
         copyFromContainerVolume(outputDir, resultingExecutableName,
                 "Failed to copy native image from container volume back to the host.");
+
+        // Note that podman cp does not support globbing i.e. cp /project/*.so will not work.
+        // Why only .so? How about .dynlib and .lib? Regardless of the host platform,
+        // the builder container is always Linux. So, we only need to copy .so files.
+        //
+        // We could either start the container again, exec `find' or `ls' to list the .so files,
+        // stop the container again and use that list. We could also use the build-artifacts.json
+        // to get the list of artifacts straight away which is what ended up doing here:
+        copyFromContainerVolume(outputDir, "build-artifacts.json", null);
+        try {
+            final Path buildArtifactsFile = outputDir.resolve("build-artifacts.json");
+            if (Files.exists(buildArtifactsFile)) {
+                // The file is small enough to afford this read
+                final String buildArtifactsJson = Files.readString(buildArtifactsFile);
+                final Matcher m = JDK_LIBS.matcher(buildArtifactsJson);
+                if (m.matches()) {
+                    for (String lib : m.group(1).split(",")) {
+                        final String l = lib.trim().replace("\"", "");
+                        copyFromContainerVolume(outputDir, l,
+                                "Failed to copy " + l + " from container volume back to the host.");
+                    }
+                }
```

Comment:
You may also use the JsonReader here, e.g.:

```suggestion
                final JsonObject jsonRead = JsonReader.of(buildArtifactsJson).read();
                JsonValue jdkLibraries = jsonRead.get("jdk_libraries");
                for (JsonValue lib : ((JsonArray) jdkLibraries).value()) {
                    copyFromContainerVolume(outputDir, ((JsonString) lib).value(),
                            "Failed to copy " + lib + " from container volume back to the host.");
                }
```

Reply:
@zakkak O.K., seemed to load way more stuff than needed for such a simple thing, but you are right. This is cleaner, it's JSON so let's read JSON...